docs/proposals/pod-safety.md

```diff
@@ -0,0 +1,431 @@
+* Once a pod begins termination, its termination grace period can only
+  be shortened, not lengthened.
+
+Pod termination is divided into the following steps:
+
+* A component requests the termination of the pod by issuing a DELETE
+  to the pod resource with an optional **grace period**
+  * If no grace period is provided, the default from the pod is leveraged
+* When the kubelet observes the deletion, it starts a timer equal to the
+  grace period and performs the following actions:
+  * Executes the pre-stop hook, if specified, waiting up to **grace period**
+    seconds before continuing
+  * Sends the termination signal to the container runtime (SIGTERM or the
+    container image's STOPSIGNAL on Docker)
+  * Waits 2 seconds, or the remaining grace period, whichever is longer
+  * Sends the force termination signal to the container runtime (SIGKILL)
+* Once the kubelet observes the container is fully terminated, it issues
+  a status update to the REST API for the pod indicating termination, then
+  issues a DELETE with grace period = 0.
+
+If the kubelet crashes during the termination process, it will restart the
+termination process from the beginning (grace period is reset). This ensures
+that a process is always given **at least** grace period to terminate cleanly.
+
+A user may re-issue a DELETE to the pod resource specifying a shorter grace
+period, but never a longer one.
+
+Deleting a pod with grace period 0 is called **force deletion** and will
+update the pod with a `deletionGracePeriodSeconds` of 0, and then immediately
+remove the pod from etcd. Because all communication is asynchronous,
+force deleting a pod means that the pod processes may continue
+to run for an arbitary amount of time. If a higher level component like the
+PetSet controller treats the existence of the pod API object as a strongly
+consistent entity, deleting the pod in this fashion will violate the
+at-most-one guarantee we wish to offer for pet sets.
+
+
+### Guarantees provided by replica sets and replication controllers
+
+ReplicaSets and ReplicationControllers both attempt to **preserve availability**
+of their constituent pods over ensuring at most one (of a pod) semantics. So a
+replica set to scale 1 will immediately create a new pod when it observes an
+old pod has begun graceful deletion, and as a result at many points in the
+lifetime of a replica set there will be 2 copies of a pod's processes running
+concurrently. Only access to exclusive resources like storage can prevent that
+simultaneous execution.
+
+Deployments, being based on replica sets, can offer no stronger guarantee.
+
+
+### Concurrent access guarantees for shared storage
+
+A persistent volume that references a strongly consistent storage backend
+like AWS EBS, GCE PD, OpenStack Cinder, or Ceph RBD can rely on the storage
+API to prevent corruption of the data due to simultaneous access by multiple
+clients. However, many commonly deployed storage technologies in the
```

Comment:
They are broadly deployed technologies that benefit from centralized control (because they offer no centralized guarantees themselves).  Most production workloads in the world today run on iSCSI, FibreChannel, or NFS devices in clustered forms - but everyone builds their own solutions to handle clustering (Pacemaker, etc).  Building the affordances in kube to allow them to be safely used is important, even if Kube itself may not contain the functionality to make them safe.